### ISSUE TITLE
References in catalogs, profiles and controls

### ISSUE THREAD

**Issue by wendellpiez** [usnistgov/OSCAL#166]
# User Story:

As an OSCAL profile designer, I wish to manage my references or external citations to arbitrary documents. I may wish to add or remove references to controls or profiles. I also expect a certain amount of behavior to be implicit: for example, when I profile a profile or baseline, I expect that all the references it makes from (in) its (selected) controls will come with those controls, while other references in the catalogs (not cited in controls) do *not* "bubble up" in that way.

## Goals:

Sketching out a robust processing model for references in profile resolution. Confirm that the present model supports it and extend the model if/as necessary. Document it so that users know what support is there and how to use it.

*NOT A GOAL*: Achieving *bibliographic description* or control over the citations as citations (for example, by providing inline markup). This is a big enough problem without trying to address that huge problem. Instead, we will look to DOIs or other machinable identifiers to help with that problem.

## Dependencies:

Realistic test cases, preferable small enough to serve as good unit tests. This probably means some kind of requirements gathering. Possibly also a decision whether to implement changes in the current ("scaffold") XSDs or the metaschema.

Also, there are closely related issues, such as error/warning support in profile resolution (not just for references), which can be addressed with this issue or at least in tandem.

## Acceptance Criteria

* We can review our test cases and confirm that they cover realistic usage
* We can confirm that they work (test profiles resolve as expected) by running them
* Our user documentation describes references with examples, including how to add or change references in profiles

**Comment by anweiss** [comment]
@wendellpiez can we maybe expand upon this issue to include better options for reference and link metadata?

The current schema allows for `prose` within a `<ref>` element, but that prose is limited to `<p>`, `<pre>`, `<ol>` and `<ul>`, none of which are good options for representing a `<ref>`'s metadata.  

Here's an example where this would be applicable. Below is a raw XML excerpt from a DISA Control Correlation Identifier (CCI) Item:

```xml
<reference creator="NIST" title="NIST SP 800-53 Revision 4" version="4" location="http://csrc.nist.gov/publications/PubsSPs.html" index="AC-1 b 1" />
```

Modeling this in its OSCAL equivalent might look something like this:

```xml
<ref id="ref001">
    <ul class="ref-data">
        <li class="creator">NIST</li>
        <li class="title">NIST SP 800-53 Revision 4</li>
        <li class="version">4</li>
        <li class="index">AC-1 b 1</li>
    </ul>
    <citation href="http://csrc.nist.gov/publications/PubsSPs.html"/>
</ref>
```

As you can see, I used a `<ul>` element with `class` attributes that point back to the original data elements and that which would be defined in separate declarations document. I could've also used `<p>` elements (with class attributes), but neither of these `prose` options is ideal for representing reference metadata.

Interestingly enough, this particular example also exposes an interesting conundrum. The original CCI example is probably better represented as a `<link>` to an external catalog as such:

```xml
<link rel="reference" href="https://raw.githubusercontent.com/usnistgov/OSCAL/master/content/nist.gov/SP800-53/rev4/NIST_SP-800-53_rev4_catalog.xml#ac-1_smt.b.1"/>`
```

But `<link>`'s also have no good ways to describe metadata per the current schema. Because of this, to preserve the original  representation of the raw data, one would also probably need to denote it as a `<ref>` in its OSCAL equivalent (as I've done in the example at the top). So we end up with a redundant representation of the same "linking"/"reference" data.

**Comment by wendellpiez** [comment]
@anweiss For reasons including, but not limited to, the issue you uncover here (about the link also being a reference) -- there's turtles here all the way down -- I want to go very slowly on mandating particular approaches to marking up and tagging references. That's why the model so far is as permissive as it is -- permitting loose prose, as reference tagging "in the wild" (ping me for references to prior art including TEI, NLM/NISO JATS, and NISO STS) has to be quite permissive in order to work at all (at least in scenarios as open-ended as ours).

This doesn't mean I don't think it's a good idea -- modeling and tagging references with more intelligence is something that might well be done in OSCAL applications. The question is, whether they could be able to exchange such data without first negotiating the details of the interface (tagging profile and constraints thereto applied). Given the history of this problem, I don't think a standard model will emerge until we have a way of automating the validation of such fields against an authoritative record (it's kind of the catalog authorization problem in miniature and two orders of magnitude larger since any catalog may have hundreds of references). And this requires authoritative endpoints, which means DOIs (digital object identifiers) with machine readable records coming back. We *almost* have those.

Over-engineering is the biggest problem with markup-based systems historically and I'd like to avoid that. 😁 This is why the ticket as written was so focused on actual data in use. Indeed, its initial focus is *not* on the potentials of bibliographic description of cited (referenced) materials, but rather on the core functionalities -- when you write a profile, how are the references in the original catalog to be handled? (Presumably, references that appear with controls, travel with the controls even if/as modified, and new references can be added. Details tbd.)

**Comment by david-waltermire-nist** [comment]
Let's review what has been done and determine if any additional work is needed.

**Comment by iMichaela** [comment]
## 5/23/2019
@david-waltermire-nist @wendellpiez and @brianrufgsa need to discuss this issue as part of @david-waltermire-nist 's review.

**Comment by david-waltermire-nist** [comment]
Worked with @wendellpiez today on cleaning up the reference, citation, and link model elements in our metaschemas. This is resulting in some content changes, which we are working through. We also removed some unused features in the catalog model, and moved some common elements to the metamodel metaschema to allow the documentation to be more broadly shared between OSCAL layers. Once done, we will write a summary of the changes we made.
